wire-ios-data-model/Source/MLS/CoreCryptoProvider.swift

```diff
@@ -0,0 +1,189 @@
+////
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+// sourcery: AutoMockable
+public protocol CoreCryptoProviderProtocol {
+
+    func coreCrypto(requireMLS: Bool) throws -> SafeCoreCryptoProtocol
+
+}
+
+public class CoreCryptoProvider: CoreCryptoProviderProtocol {
+    private let selfUserID: UUID
+    private let sharedContainerURL: URL
+    private let accountDirectory: URL
+    private let cryptoboxMigrationManager: CryptoboxMigrationManagerInterface
+    private let syncContext: NSManagedObjectContext
+    private let lock = NSLock()
+    private var coreCrypto: SafeCoreCrypto?
+
+    public init(selfUserID: UUID,
+                sharedContainerURL: URL,
+                accountDirectory: URL,
+                syncContext: NSManagedObjectContext,
+                cryptoboxMigrationManager: CryptoboxMigrationManagerInterface) {
+        self.selfUserID = selfUserID
+        self.sharedContainerURL = sharedContainerURL
+        self.accountDirectory = accountDirectory
+        self.syncContext = syncContext
+        self.cryptoboxMigrationManager = cryptoboxMigrationManager
+    }
+
+    // NOTE: this will turn async when we upgrade CC
+    public func coreCrypto(requireMLS: Bool = false) throws -> SafeCoreCryptoProtocol {
+        // TODO: this lock should go away when this function turn async and the class can become an actor
+        return try lock.withLock {
+            let coreCrypto = if let coreCrypto = coreCrypto {
+                coreCrypto
+            } else {
+                try createCoreCrypto()
+            }
+
+            self.coreCrypto = coreCrypto
+
+            if requireMLS {
+                let provider = CoreCryptoConfigProvider()
+                let clientID = try provider.clientID(of: .selfUser(in: syncContext))
```

Comment:
> I see we get the selfuser later for the client keys what about holding to the selfuser ?

I can't inject the clientID on init since it might not exist then.